Add contents for CppInterOp Documentation (#105)
compiler-research/CppInterOp: docs/tutorials.rst

> Tutorials

Reviewer: @wlav, could you take a look and give feedback on our documentation intent to capture the cppyy/python interop use case as motivation for CppInterOp?

Reviewer 2: ? What specific piece of text needs commenting?

Reviewer: @wlav, basically this file: `docs/tutorials.rst`. I'd like to make sure we are not making some unreasonable claim.

----------
This tutorial emphasizes on the abilities and usage of CppInterOp, Let's get started!
The tutorial demonstrates two examples one in C and python for interoperability.

> **Note:: This example library shown below is portrayal of the concept on which 

Reviewer: ```suggestion
**Note:: This example library shown below is illustrate the concept on which 
```

Tutorials
----------
This tutorial emphasizes on the abilities and usage of CppInterOp, Let's get
started! The tutorial demonstrates two examples one in C and python for
interoperability.

**Note:: This example library shown below is to illustrate the concept on which
CppInterOp is based on.** 

Python:
=======

.. code-block:: bash

   libInterop = ctypes.CDLL(libpath, mode = ctypes.RTLD_GLOBAL)
   _cpp_compile = libInterop.Clang_Parse
   _cpp_compile.argtypes = [ctypes.c_char_p]

> We are using ctypes for inducting our library, *Clang_Parse* which is part

Reviewer: Where are Clang_Parse and such defined? Should we not show how the header file looks like at least?

    Decl_t Clang_LookupName(const char* Name, Decl_t Context);

    # Returns the address of a JIT'd function of the corresponding declaration.
    FnAddr_t Clang_GetFunctionAddress(Decl_t D);

    # Returns the name of any named decl (class, namespace) & template arguments
    std::string GetCompleteName(Decl_t A);

    # Allocates memory of underlying size of the passed declaration.
    void * Clang_CreateObject(Decl_t RecordDecl);

    # Instantiates a given templated declaration.
    Decl_t Clang_InstantiateTemplate(Decl_t D, const char* Name, const char* Args);

.. code-block:: cpp

   def cpp_compile(arg):
    return _cpp_compile(arg.encode("ascii"))

   cpp_compile(r"""\
   void* operator new(__SIZE_TYPE__, void* __p) noexcept;
   extern "C" int printf(const char*,...);
   class A {};
   class C {};
   struct B : public A {
   template<typename T, typename S, typename U>
   void callme(T, S, U*) { printf(" call me may B! \n"); }
   };
   """)

The C++ code which is to be used in python comes under this section, this code 
is parsed by the CppInterOp library in previous snippet and further compilation
goes on.

> .. code-block:: cpp

Reviewer: No, that’s python.

  def _get_template(self, scope, name, args):
    return self._get_template_ct(scope, name.encode("ascii"), args.encode("ascii"))

  def get_scope(self, name):
    return self._get_scope(name.encode("ascii"))

  def get_template(self, scope, name, tmpl_args = [], tpargs = []):
    if tmpl_args:
      # Instantiation is explicit from full name
      full_name = name + '<' + ', '.join([a for a in tmpl_args]) + '>'
      meth = self._get_template(scope, full_name, '')
    elif tpargs:
      # Instantiation is implicit from argument types
      meth = self._get_template(scope, name, ', '.join([a.__name__ for a in tpargs]))
    return CallCPPFunc(meth)

  def construct(self, cpptype):
    return self._construct(cpptype)

The class InterOpLayerWrapper is supposed to provide a python wrapper over
the cppinterop layer.In this we have the functions *Clang_LookupName,
Clang_CreateObject and Clang_InstantiateTemplate* which are being used, so these
are being wrapped to be used in the python language.

We get to know the scope of the attribute of class by get_scope which is being 
wrapped. In the similar manner we can use get_namespace.

> .. code-block::cpp

Reviewer: Likewise.

Tutorials
----------

> This tutorial emphasizes on the abilities and usage of CppInterOp, Let's get

Reviewer: ```suggestion
This tutorial emphasizes on the abilities and usage of CppInterOp. Let's get
```

Can you please reread *very carefully* the entire set of documentation. I do not expect to continue fixing such details as it seems that we have not paid enough attention at that point. Please make sure it does not have any obvious issues before asking for a review.

   _cpp_compile = libInterop.Clang_Parse
   _cpp_compile.argtypes = [ctypes.c_char_p]

  # We are using ctypes for inducting our library, *Clang_Parse* which is part
  # of the library is being used for parsing the C++ code.

> .. code-block:: bash

Reviewer: ```suggestion
.. code-block:: cpp
```

Then replace # with //
